external-crates/move/crates/move-compiler/src/sui_mode/linters/missing_key.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This linter rule checks for structs with an `id` field of type `UID` without the `key` ability.
+
+use super::{LinterDiagnosticCategory, LinterDiagnosticCode, LINT_WARNING_PREFIX};
+use crate::expansion::ast::ModuleIdent;
+use crate::parser::ast::DatatypeName;
+use crate::{
+    diag,
+    diagnostics::{
+        codes::{custom, DiagnosticInfo, Severity},
+        WarningFilters,
+    },
+    naming::ast::{StructDefinition, StructFields},
+    parser::ast::Ability_,
+    shared::CompilationEnv,
+    typing::{
+        ast as T,
+        visitor::{TypingVisitorConstructor, TypingVisitorContext},
+    },
+};
+use move_ir_types::location::Loc;
+
+const MISSING_KEY_ABILITY_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagnosticCategory::Sui as u8,
+    LinterDiagnosticCode::MissingKey as u8,
+    "The struct's first field is 'id' of type 'UID' but is missing the 'key' ability.",
+);
+
+pub struct MissingKeyVisitor;
+
+pub struct Context<'a> {
+    env: &'a mut CompilationEnv,
+}
+impl TypingVisitorConstructor for MissingKeyVisitor {
+    type Context<'a> = Context<'a>;
+
+    fn context<'a>(env: &'a mut CompilationEnv, _program: &T::Program) -> Self::Context<'a> {
+        Context { env }
+    }
+}
+
+impl TypingVisitorContext for Context<'_> {
+    fn add_warning_filter_scope(&mut self, filter: WarningFilters) {
+        self.env.add_warning_filter_scope(filter)
+    }
+
+    fn pop_warning_filter_scope(&mut self) {
+        self.env.pop_warning_filter_scope()
+    }
+
+    const VISIT_TYPES: bool = true;
+
```

Comment:
```suggestion
```
This isn't needed for this visitor

Reply:
The visitor not running if `VISIT_TYPES = false`

Reply:
Ah, thanks for the heads up. That is not the intended behavior. We will fix that after this lands